File: agent/loop.py
from __future__ import annotations

import logging
from pathlib import Path

from ..config import Config
from ..permissions import PermissionPolicy
from ..providers import get_provider
from ..providers.base import tool_results_as_message
from ..tools import Tool, execute_parallel, build_registry
from ..tools.files import FILE_TOOLS
from ..tools.shell import SHELL_TOOLS
from ..tools.web import WEB_TOOLS
from . import compactor
from . import session as sess

log = logging.getLogger(__name__)

_CONTEXT_WINDOWS = {
    "claude-opus": 200_000,
    "claude-sonnet": 200_000,
    "claude-haiku": 200_000,
    "gpt-4": 128_000,
    "gpt-5.5-pro": 128_000,
    "default": 128_000,
}


def _context_window(model: str) -> int:
    for prefix, size in _CONTEXT_WINDOWS.items():
        if prefix in model:
            return size
    return _CONTEXT_WINDOWS["default"]


def default_tools() -> list[Tool]:
    return FILE_TOOLS + SHELL_TOOLS + WEB_TOOLS


async def run(
    messages: list[dict],
    config: Config,
    *,
    extra_tools: list[Tool] | None = None,
    session_id: str | None = None,
    policy: PermissionPolicy | None = None,
) -> str:
    provider = get_provider(config)
    tools = default_tools() + (extra_tools or [])
    registry = build_registry(tools)
    api_tools = [t.as_api_dict() for t in tools]
    policy = policy or PermissionPolicy.from_mode(config.permission_mode)
    ctx_window = _context_window(config.model)

    for iteration in range(config.max_iterations):
        # compact if approaching context limit
        if compactor.needs_compaction(messages, ctx_window):
            messages = await compactor.compact(messages, provider, config.model)

        response = await provider.stream(messages, api_tools, config.model)
        messages.append(response.as_assistant_message())

        if session_id:
            sess.append_turn(config.sessions_dir, session_id, [response.as_assistant_message()])

        log.debug(
            "iter=%d stop=%s tokens_in=%d tokens_out=%d",
            iteration, response.stop_reason, response.input_tokens, response.output_tokens,
        )

        if response.stop_reason != "tool_use":
            return response.text

        if response.input_tokens + response.output_tokens >= config.max_tokens:
            log.warning("Token limit reached at iteration %d", iteration)
            return response.text or "[max tokens reached]"

        results = await execute_parallel(response.tool_uses, registry, policy)
        tool_msg = tool_results_as_message(results)
        messages.append(tool_msg)

        if session_id:
            sess.append_turn(config.sessions_dir, session_id, [tool_msg])

    return "[max iterations reached]"

File: src/agent/loop.ts
import type { Config } from "../config.ts";
import type { Message } from "../providers/base.ts";
import { toolResultsAsMessage } from "../providers/base.ts";
import { getProvider } from "../providers/index.ts";
import { PermissionPolicy } from "../permissions.ts";
import type { Tool } from "../tools/index.ts";
import { executeParallel, buildRegistry } from "../tools/index.ts";
import { FILE_TOOLS } from "../tools/files.ts";
import { SHELL_TOOLS } from "../tools/shell.ts";
import { WEB_TOOLS } from "../tools/web.ts";
import { needsCompaction, compact } from "./compactor.ts";
import { appendTurn } from "./session.ts";

const CONTEXT_WINDOWS: Record<string, number> = {
  "claude-opus": 200_000,
  "claude-sonnet": 200_000,
  "claude-haiku": 200_000,
  "gpt-4": 128_000,
  "gpt-5": 128_000,
  default: 128_000,
};

function contextWindow(model: string): number {
  for (const [prefix, size] of Object.entries(CONTEXT_WINDOWS)) {
    if (model.includes(prefix)) return size;
  }
  return CONTEXT_WINDOWS["default"]!;
}

export function defaultTools(): Tool[] {
  return [...FILE_TOOLS, ...SHELL_TOOLS, ...WEB_TOOLS];
}

export interface RunOptions {
  extraTools?: Tool[];
  sessionId?: string;
  policy?: PermissionPolicy;
}

export async function run(
  messages: Message[],
  config: Config,
  opts?: RunOptions,
): Promise<string> {
  const provider = getProvider(config);
  const tools = [...defaultTools(), ...(opts?.extraTools ?? [])];
  const registry = buildRegistry(tools);
  const apiTools = tools.map((t) => t.asApiDict());
  const policy = opts?.policy ?? PermissionPolicy.fromMode(config.permissionMode);
  const ctxWindow = contextWindow(config.model);

  for (let i = 0; i < config.maxIterations; i++) {
    // compact if approaching context limit
    if (needsCompaction(messages, ctxWindow)) {
      messages = await compact(messages, provider, config.model);
    }

    const response = await provider.stream(messages, apiTools, config.model);
    const assistantMsg = response.asAssistantMessage();
    messages.push(assistantMsg);

    if (opts?.sessionId) {
      await appendTurn(config.sessionsDir, opts.sessionId, [assistantMsg]);
    }

    console.debug(
      `iter=${i} stop=${response.stopReason} tokens_in=${response.inputTokens} tokens_out=${response.outputTokens}`,
    );

    if (response.stopReason !== "tool_use") {
      return response.text;
    }

    if (response.inputTokens + response.outputTokens >= config.maxTokens) {
      console.warn(`Token limit reached at iteration ${i}`);
      return response.text || "[max tokens reached]";
    }

    const results = await executeParallel(response.toolUses, registry, policy);
    const toolMsg = toolResultsAsMessage(results);
    messages.push(toolMsg);

    if (opts?.sessionId) {
      await appendTurn(config.sessionsDir, opts.sessionId, [toolMsg]);
    }
  }

  return "[max iterations reached]";
}
